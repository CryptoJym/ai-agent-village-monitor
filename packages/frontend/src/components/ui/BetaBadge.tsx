import React from 'react';
import { useFeatureFlags } from '../../contexts/FeatureFlags';

interface BetaBadgeProps {
  position?: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right' | 'inline';
  size?: 'small' | 'medium' | 'large';
  variant?: 'primary' | 'gradient' | 'minimal';
  showLabel?: boolean;
  onClick?: () => void;
}

export function BetaBadge({
  position = 'top-right',
  size = 'medium',
  variant = 'gradient',
  showLabel = true,
  onClick,
}: BetaBadgeProps) {
  const { isBetaTester } = useFeatureFlags();

  // Don't render if user is not a beta tester
  if (!isBetaTester) return null;

  const sizeStyles = {
    small: {
      padding: '4px 8px',
      fontSize: '10px',
      borderRadius: '8px',
    },
    medium: {
      padding: '6px 12px',
      fontSize: '12px',
      borderRadius: '10px',
    },
    large: {
      padding: '8px 16px',
      fontSize: '14px',
      borderRadius: '12px',
    },
  };

  const variantStyles = {
    primary: {
      background: '#3b82f6',
      color: '#ffffff',
      border: '1px solid #2563eb',
    },
    gradient: {
      background: 'linear-gradient(45deg, #3b82f6, #8b5cf6)',
      color: '#ffffff',
      border: 'none',
    },
    minimal: {
      background: 'rgba(59, 130, 246, 0.1)',
      color: '#60a5fa',
      border: '1px solid rgba(59, 130, 246, 0.3)',
    },
  };

  const positionStyles = {
    'top-left': {
      position: 'fixed' as const,
      top: '16px',
      left: '16px',
      zIndex: 1000,
    },
    'top-right': {
      position: 'fixed' as const,
      top: '16px',
      right: '16px',
      zIndex: 1000,
    },
    'bottom-left': {
      position: 'fixed' as const,
      bottom: '16px',
      left: '16px',
      zIndex: 1000,
    },
    'bottom-right': {
      position: 'fixed' as const,
      bottom: '16px',
      right: '16px',
      zIndex: 1000,
    },
    inline: {
      position: 'relative' as const,
      display: 'inline-block',
    },
  };

  const badgeStyles: React.CSSProperties = {
    ...sizeStyles[size],
    ...variantStyles[variant],
    ...positionStyles[position],
    fontWeight: 700,
    textTransform: 'uppercase' as const,
    letterSpacing: '0.5px',
    cursor: onClick ? 'pointer' : 'default',
    transition: 'all 0.3s ease',
    boxShadow: variant === 'gradient' ? '0 4px 12px rgba(59, 130, 246, 0.3)' : 'none',
    userSelect: 'none' as const,
    fontFamily: 'system-ui, sans-serif',
  };

  const handleClick = () => {
    if (onClick) {
      onClick();
    }
  };

  const handleMouseEnter = (e: React.MouseEvent<HTMLDivElement>) => {
    if (onClick) {
      e.currentTarget.style.transform = 'translateY(-2px)';
      e.currentTarget.style.boxShadow = '0 6px 16px rgba(59, 130, 246, 0.4)';
    }
  };

  const handleMouseLeave = (e: React.MouseEvent<HTMLDivElement>) => {
    if (onClick) {
      e.currentTarget.style.transform = 'translateY(0)';
      e.currentTarget.style.boxShadow = variant === 'gradient' ? '0 4px 12px rgba(59, 130, 246, 0.3)' : 'none';
    }
  };

  return (
    <div
      style={badgeStyles}
      onClick={handleClick}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      role={onClick ? 'button' : 'img'}
      aria-label={showLabel ? 'Beta version indicator' : undefined}
      tabIndex={onClick ? 0 : undefined}
      onKeyDown={onClick ? (e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          e.preventDefault();
          handleClick();
        }
      } : undefined}
    >
      {showLabel ? 'Beta' : 'Î²'}
    </div>
  );
}

// Specialized version for header/navigation areas
export function HeaderBetaBadge() {
  return (
    <BetaBadge
      position="inline"
      size="small"
      variant="minimal"
      showLabel={true}
    />
  );
}

// Floating beta badge with feedback action
export function FloatingBetaBadge({ onFeedbackClick }: { onFeedbackClick?: () => void }) {
  return (
    <BetaBadge
      position="bottom-right"
      size="medium"
      variant="gradient"
      showLabel={true}
      onClick={onFeedbackClick}
    />
  );
}