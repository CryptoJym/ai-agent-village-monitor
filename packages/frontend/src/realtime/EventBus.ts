import mitt from 'mitt';

type Events = {
  agent_update: { agentId: string; state: string; x?: number; y?: number };
  work_stream: { agentId: string; message: string; ts?: number };
  // Spawns may reference a house (preferred) or provide absolute x/y as a fallback.
  // When houseId is provided, the scene will place the bot within a ring around the house center.
  bug_bot_spawn: {
    id: string;
    // Optional target house context
    houseId?: string;
    title?: string;
    summary?: string;
    // Fallback coordinates (e.g., house center or absolute spawn)
    x?: number;
    y?: number;
    severity?: 'low' | 'medium' | 'high';
  };
  bug_bot_progress: { id: string; progress: number }; // 0..1
  bug_bot_resolved: { id: string };
  bug_bot_assign_request: { id: string };
  agent_drop: { x: number; y: number };
  agent_assignment: { agentId: string; houseId?: string };
  agent_identity: { agentId: string; name?: string };
  house_focus: { houseId: string; source?: string };
  house_dashboard_request: { houseId: string; source?: string };
  house_dashboard: {
    houseId: string;
    name: string;
    language?: string;
    components?: string[];
    issues?: number;
    agents?: Array<{ id: string; name: string }>;
    stars?: number;
    buildStatus?: string;
  };
  connection_status: { status: 'connecting' | 'connected' | 'disconnected' };
  latency: { rttMs: number };
  toast: { type: 'success' | 'error' | 'info'; message: string };
  // Emitted when the main camera finishes an instant snap or a short pan
  cameraSettled: { x: number; y: number; zoom: number };
  house_activity: {
    type: 'house.activity';
    houseId?: string | number;
    repoId?: string | number;
    indicators: {
      lights: { active: boolean; minRemainingMs?: number };
      banner: { active: boolean; prNumber?: number; minRemainingMs?: number };
      smoke: {
        active: boolean;
        status?: 'in_progress' | 'failed' | 'passed';
        minRemainingMs?: number;
      };
    };
    version: number;
    ts: number;
  };
};

export const eventBus = mitt<Events>();
