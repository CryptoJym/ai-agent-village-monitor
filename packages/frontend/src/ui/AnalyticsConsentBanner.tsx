import { useEffect, useState } from 'react';
import { hasStoredConsent, setConsent } from '../analytics/client';

export function AnalyticsConsentBanner() {
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    try {
      const nav: any = typeof navigator !== 'undefined' ? navigator : undefined;
      const gpc = !!(nav && 'globalPrivacyControl' in nav && nav.globalPrivacyControl);
      const dnt =
        nav?.doNotTrack === '1' ||
        nav?.msDoNotTrack === '1' ||
        (typeof window !== 'undefined' && (window as any)?.doNotTrack === '1');
      if (!hasStoredConsent() && !gpc && !dnt) setVisible(true);
    } catch (e) {
      void e;
    }
  }, []);

  if (!visible) return null;

  const choose = async (allow: boolean) => {
    setConsent(allow);
    setVisible(false);
    try {
      // Persist on server if logged in
      await fetch('/api/users/me/preferences', {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        credentials: 'include',
        body: JSON.stringify({ analytics: { enabled: allow } }),
      });
    } catch (e) {
      void e;
    }
  };

  return (
    <div
      style={{
        position: 'fixed',
        left: 16,
        bottom: 16,
        padding: '12px 14px',
        background: '#0b1220',
        color: '#e2e8f0',
        border: '1px solid #334155',
        borderRadius: 8,
        zIndex: 50,
        display: 'flex',
        gap: 12,
        alignItems: 'center',
        justifyContent: 'space-between',
        maxWidth: 'min(420px, calc(100vw - 40px))',
        boxShadow: '0 10px 30px rgba(15, 23, 42, 0.55)',
      }}
      role="region"
      aria-label="Analytics consent"
    >
      <div style={{ maxWidth: '60%' }}>
        Help us improve with optional usage analytics (dialogue opens, session duration, etc.). We only
        start collecting after you opt in. See our{' '}
        <a href="/docs/privacy" style={{ color: '#93c5fd' }}>
          privacy policy
        </a>{' '}
        for details.
      </div>
      <div style={{ display: 'flex', gap: 8 }}>
        <button
          type="button"
          onClick={() => void choose(false)}
          style={{
            padding: '8px 12px',
            background: 'transparent',
            color: '#e2e8f0',
            border: '1px solid #475569',
            borderRadius: 6,
            cursor: 'pointer',
          }}
        >
          No thanks
        </button>
        <button
          type="button"
          onClick={() => void choose(true)}
          style={{
            padding: '8px 12px',
            background: '#1f2937',
            color: '#e2e8f0',
            border: '1px solid #374151',
            borderRadius: 6,
            cursor: 'pointer',
          }}
        >
          Allow
        </button>
      </div>
    </div>
  );
}
