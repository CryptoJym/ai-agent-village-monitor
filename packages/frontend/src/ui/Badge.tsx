import React from 'react';

type Role = 'owner' | 'member' | 'visitor' | 'you';

const COLORS: Record<Role, { bg: string; fg: string; border: string; label: string }> = {
  owner: { bg: '#064e3b', fg: '#d1fae5', border: '#065f46', label: 'Owner' },
  member: { bg: '#0c4a6e', fg: '#e0f2fe', border: '#075985', label: 'Member' },
  visitor: { bg: '#111827', fg: '#e5e7eb', border: '#1f2937', label: 'Visitor' },
  you: { bg: '#1f2937', fg: '#fef08a', border: '#374151', label: 'You' },
};

export function Badge({ role, title }: { role: Role; title?: string }) {
  const c = COLORS[role];
  return (
    <span
      title={title || c.label}
      aria-label={`${c.label} badge`}
      style={{
        display: 'inline-block',
        padding: '2px 8px',
        borderRadius: 999,
        background: c.bg,
        color: c.fg,
        border: `1px solid ${c.border}`,
        fontSize: 11,
        lineHeight: '16px',
        fontFamily: 'system-ui, sans-serif',
      }}
    >
      {c.label}
    </span>
  );
}
