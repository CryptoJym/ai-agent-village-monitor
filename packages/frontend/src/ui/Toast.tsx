import React, { createContext, useCallback, useContext, useMemo, useState } from 'react';

export type Toast = { id: string; type: 'info' | 'error' | 'success'; text: string };

type ToastContextValue = {
  toasts: Toast[];
  showInfo: (text: string) => void;
  showError: (text: string) => void;
  showSuccess: (text: string) => void;
};

const ToastContext = createContext<ToastContextValue | null>(null);

export function useToast() {
  const ctx = useContext(ToastContext);
  if (!ctx) throw new Error('useToast must be used within ToastProvider');
  return ctx;
}

export function ToastProvider({ children }: { children?: React.ReactNode }) {
  const [toasts, setToasts] = useState<Toast[]>([]);

  const push = useCallback((type: Toast['type'], text: string) => {
    const id = `${Date.now()}-${Math.random().toString(36).slice(2)}`;
    const t: Toast = { id, type, text };
    setToasts((prev) => [...prev, t]);
    // auto-dismiss after 3s
    setTimeout(() => {
      setToasts((prev) => prev.filter((x) => x.id !== id));
    }, 3000);
  }, []);

  const value = useMemo<ToastContextValue>(
    () => ({
      toasts,
      showInfo: (text) => push('info', text),
      showError: (text) => push('error', text),
      showSuccess: (text) => push('success', text),
    }),
    [toasts, push],
  );

  return (
    <ToastContext.Provider value={value}>
      {children}
      <ToastViewport toasts={toasts} />
    </ToastContext.Provider>
  );
}

function ToastViewport({ toasts }: { toasts: Toast[] }) {
  if (typeof process !== 'undefined' && process.env.NODE_ENV === 'test') {
    return null;
  }
  return (
    <div
      style={{
        position: 'fixed',
        right: 12,
        top: 12,
        display: 'flex',
        flexDirection: 'column',
        gap: 8,
        zIndex: 50,
      }}
    >
      {toasts.map((t) => (
        <div
          key={t.id}
          role="status"
          aria-live="polite"
          style={{
            padding: '8px 12px',
            borderRadius: 8,
            border: '1px solid #1f2937',
            background:
              t.type === 'error' ? '#7f1d1d' : t.type === 'success' ? '#064e3b' : '#1f2937',
            color: '#e5e7eb',
            minWidth: 240,
            boxShadow: '0 6px 24px rgba(0,0,0,0.25)',
          }}
        >
          {t.text}
        </div>
      ))}
    </div>
  );
}
