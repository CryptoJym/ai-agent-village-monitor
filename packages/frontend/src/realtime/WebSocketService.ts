import { io, Socket } from 'socket.io-client';
import { eventBus } from './EventBus';

export type WebSocketOptions = {
  url?: string; // default derived from location
  token?: string; // JWT
  transports?: ('websocket' | 'polling')[];
  reconnectionAttempts?: number;
  reconnectionDelay?: number;
};

export class WebSocketService {
  private socket?: Socket;
  private opts: Required<WebSocketOptions>;
  // Throttled event buffers
  private wsQueue: any[] = [];
  private rafScheduled = false;

  constructor(options: WebSocketOptions = {}) {
    const envWs = (import.meta as any)?.env?.VITE_WS_URL as string | undefined;
    const defaultUrl =
      envWs ??
      (typeof location !== 'undefined'
        ? `${location.protocol === 'https:' ? 'wss' : 'ws'}://${location.hostname}:3000`
        : 'ws://localhost:3000');
    this.opts = {
      url: options.url ?? defaultUrl,
      token: options.token ?? '',
      // Allow polling fallback; Socket.IO will prefer websocket when available
      transports: options.transports ?? ['websocket', 'polling'],
      reconnectionAttempts: options.reconnectionAttempts ?? 10,
      reconnectionDelay: options.reconnectionDelay ?? 1000,
    };
  }

  connect() {
    if (this.socket) return;
    eventBus.emit('connection_status', { status: 'connecting' });
    this.socket = io(this.opts.url, {
      transports: this.opts.transports,
      autoConnect: true,
      reconnection: true,
      reconnectionAttempts: this.opts.reconnectionAttempts,
      reconnectionDelay: this.opts.reconnectionDelay,
      auth: (cb) => cb({ token: this.opts.token }),
    });

    this.socket.on('connect', () => {
      eventBus.emit('connection_status', { status: 'connected' });
      // Start latency measurements using socket.io's own ping
      const start = performance.now();
      this.socket?.timeout(2000).emit('ping', () => {
        const rttMs = performance.now() - start;
        eventBus.emit('latency', { rttMs });
      });
      // Optional catch-up flow after reconnect
      void this.fetchCatchup().catch(() => {});
    });

    this.socket.on('disconnect', () => {
      eventBus.emit('connection_status', { status: 'disconnected' });
    });

    // Server events â†’ event bus (throttled for high-frequency streams)
    this.socket.on('agent_update', (payload) => this.enqueue('agent_update', payload));
    this.socket.on('work_stream', (payload) => this.enqueue('work_stream', payload));
    this.socket.on('bug_bot_spawn', (payload) => eventBus.emit('bug_bot_spawn', payload));
    this.socket.on('bug_bot_progress', (payload) => eventBus.emit('bug_bot_progress', payload));
    this.socket.on('bug_bot_resolved', (payload) => eventBus.emit('bug_bot_resolved', payload));
  }

  disconnect() {
    this.socket?.disconnect();
    this.socket = undefined;
  }

  private enqueue(type: 'agent_update' | 'work_stream', payload: any) {
    // In Vitest, emit synchronously to simplify tests
    try {
      const inTest =
        typeof process !== 'undefined' &&
        ((process as any).env?.VITEST || (process as any).env?.VITEST_WORKER_ID);
      if (inTest) {
        eventBus.emit(type, payload);
        return;
      }
    } catch {}
    this.wsQueue.push({ type, payload });
    if (!this.rafScheduled) {
      this.rafScheduled = true;
      const schedule =
        typeof requestAnimationFrame === 'function'
          ? requestAnimationFrame
          : (cb: FrameRequestCallback) => setTimeout(cb, 16) as unknown as number;
      schedule(() => this.flush());
    }
  }

  private flush() {
    this.rafScheduled = false;
    if (this.wsQueue.length === 0) return;
    const q = this.wsQueue.splice(0);
    // Coalesce by type if needed; for now, emit in order
    for (const e of q) eventBus.emit(e.type, e.payload);
  }

  joinVillage(villageId: string) {
    this.socket?.emit('join_village', { villageId });
  }

  joinAgent(agentId: string) {
    this.socket?.emit('join_agent', { agentId });
  }

  // Placeholder REST catch-up; adapt to your backend API
  async fetchCatchup(): Promise<void> {
    try {
      // e.g., const res = await fetch('/api/catchup'); await res.json();
      return;
    } catch {
      return;
    }
  }
}

// Convenience singleton for app-wide use if desired
export const ws = new WebSocketService();
