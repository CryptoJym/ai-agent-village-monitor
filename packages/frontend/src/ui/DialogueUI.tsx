import { useEffect, useMemo, useRef, useState } from 'react';
import { getUIState, setUIState } from '../state/uiState';
import { ThreadTab } from './ThreadTab';
import { useTranslation } from 'react-i18next';
import { ControlTab } from './ControlTab';
import { InfoTab } from './InfoTab';
import { ToastProvider } from './Toast';

export type DialogueUIProps = {
  open: boolean;
  onClose: () => void;
  agentId?: string;
  initialTab?: 'thread' | 'control' | 'info';
  onTabChange?: (tab: 'thread' | 'control' | 'info') => void;
};

type TabKey = 'thread' | 'control' | 'info';

export function DialogueUI({
  open,
  onClose,
  agentId = 'demo-agent',
  initialTab,
  onTabChange,
}: DialogueUIProps) {
  const { t } = useTranslation();
  const [tab, setTab] = useState<TabKey>(() => {
    if (initialTab) return initialTab;
    try {
      const st = getUIState();
      if (
        st.dialogueTab &&
        (st.dialogueTab === 'thread' || st.dialogueTab === 'control' || st.dialogueTab === 'info')
      ) {
        return st.dialogueTab as any;
      }
    } catch {}
    return 'thread';
  });
  const [heightPct, setHeightPct] = useState(0.3);
  const overlayRef = useRef<HTMLDivElement | null>(null);
  const closeBtnRef = useRef<HTMLButtonElement | null>(null);
  const prevFocusRef = useRef<HTMLElement | null>(null);

  useEffect(() => {
    const update = () => {
      const mobile = window.innerWidth <= 640;
      setHeightPct(mobile ? 0.5 : 0.3);
    };
    update();
    window.addEventListener('resize', update);
    return () => window.removeEventListener('resize', update);
  }, []);

  useEffect(() => {
    const onKey = (e: KeyboardEvent) => {
      if (!open) return;
      if (e.key === 'Escape') {
        onClose();
      } else if (e.key === '1') {
        e.preventDefault();
        setTab('thread');
      } else if (e.key === '2') {
        e.preventDefault();
        setTab('control');
      } else if (e.key === '3') {
        e.preventDefault();
        setTab('info');
      }
    };
    window.addEventListener('keydown', onKey);
    return () => window.removeEventListener('keydown', onKey);
  }, [open, onClose]);

  // Persist selected agent and tab; restore tab on open
  useEffect(() => {
    try {
      setUIState({ selectedAgentId: agentId });
    } catch {}
  }, [agentId]);

  useEffect(() => {
    if (open) {
      try {
        const st = getUIState();
        if (
          st.dialogueTab &&
          (st.dialogueTab === 'thread' || st.dialogueTab === 'control' || st.dialogueTab === 'info')
        ) {
          setTab(st.dialogueTab as any);
        }
      } catch {}
    }
  }, [open]);

  useEffect(() => {
    try {
      setUIState({ dialogueTab: tab as any });
    } catch {}
  }, [tab]);

  // Reflect prop changes to initialTab (e.g., hash hydration)
  useEffect(() => {
    if (initialTab && initialTab !== tab) setTab(initialTab);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [initialTab]);

  // Notify parent of tab changes (for URL hash sync)
  useEffect(() => {
    onTabChange?.(tab);
  }, [tab, onTabChange]);

  // Focus management: trap initial focus and restore on close
  useEffect(() => {
    if (open) {
      prevFocusRef.current = (document.activeElement as HTMLElement) || null;
      setTimeout(() => closeBtnRef.current?.focus(), 0);
    } else {
      // restore focus to previous element
      prevFocusRef.current?.focus?.();
    }
  }, [open]);

  const styles = useMemo((): {
    overlay: React.CSSProperties;
    panel: React.CSSProperties;
    tabs: React.CSSProperties;
    tabBtn: (active: boolean) => React.CSSProperties;
    body: React.CSSProperties;
  } => {
    const panelH = `${Math.round(heightPct * 100)}vh`;
    return {
      overlay: {
        position: 'fixed' as const,
        inset: 0,
        background: open ? 'rgba(0,0,0,0.35)' : 'transparent',
        pointerEvents: (open ? 'auto' : 'none') as React.CSSProperties['pointerEvents'],
        transition: 'background 200ms ease',
        zIndex: 30,
      },
      panel: {
        position: 'absolute' as const,
        left: 0,
        right: 0,
        bottom: 0,
        height: panelH,
        background: '#0b1220',
        borderTop: '1px solid #1f2937',
        boxShadow: '0 -10px 30px rgba(0,0,0,0.25)',
        transform: open ? 'translateY(0%)' : 'translateY(100%)',
        transition: 'transform 300ms ease-out',
        display: 'flex',
        flexDirection: 'column' as const,
      },
      tabs: {
        display: 'flex',
        gap: 8,
        padding: '8px 12px',
        borderBottom: '1px solid #1f2937',
        background: '#0f172a',
      },
      tabBtn: (active: boolean) => ({
        padding: '6px 10px',
        borderRadius: 6,
        background: active ? '#1f2937' : 'transparent',
        color: '#e5e7eb',
        border: '1px solid #374151',
        cursor: 'pointer',
      }),
      body: {
        flex: 1,
        minHeight: 0,
        overflow: 'hidden',
      },
    };
  }, [open, heightPct]);

  const stop = (e: React.MouseEvent) => e.stopPropagation();

  return (
    <div
      ref={overlayRef}
      style={styles.overlay}
      onClick={onClose}
      data-testid="dialogue-overlay"
      aria-hidden={!open}
    >
      <section
        style={styles.panel}
        onClick={stop}
        data-testid="dialogue-panel"
        role="dialog"
        aria-modal="true"
        aria-label={t('dialogue.ariaPanel')}
      >
        <header style={styles.tabs} data-testid="dialogue-tabs" aria-label={t('dialogue.ariaTabs')}>
          <button
            style={styles.tabBtn(tab === 'thread')}
            onClick={() => setTab('thread')}
            aria-pressed={tab === 'thread'}
            aria-label={`${t('dialogue.thread')} tab (1)`}
          >
            {t('dialogue.thread')}
          </button>
          <button
            style={styles.tabBtn(tab === 'control')}
            onClick={() => setTab('control')}
            aria-pressed={tab === 'control'}
            aria-label={`${t('dialogue.control')} tab (2)`}
          >
            {t('dialogue.control')}
          </button>
          <button
            style={styles.tabBtn(tab === 'info')}
            onClick={() => setTab('info')}
            aria-pressed={tab === 'info'}
            aria-label={`${t('dialogue.info')} tab (3)`}
          >
            {t('dialogue.info')}
          </button>
          <div style={{ marginLeft: 'auto' }}>
            <button
              style={{
                padding: '6px 10px',
                borderRadius: 6,
                background: 'transparent',
                color: '#e5e7eb',
                border: '1px solid #374151',
                cursor: 'pointer',
              }}
              onClick={onClose}
              aria-label={t('dialogue.closeAria')}
              ref={closeBtnRef}
            >
              {t('dialogue.closeText')}
            </button>
          </div>
        </header>
        <ToastProvider>
          <div style={styles.body}>
            {tab === 'thread' && <ThreadTab agentId={agentId} />}
            {tab === 'control' && <ControlTab agentId={agentId} />}
            {tab === 'info' && <InfoTab agentId={agentId} />}
          </div>
        </ToastProvider>
      </section>
    </div>
  );
}
