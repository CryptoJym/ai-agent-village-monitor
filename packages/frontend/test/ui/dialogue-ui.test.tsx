import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import { DialogueUI } from '../../src/ui/DialogueUI';
import { ToastProvider } from '../../src/ui/Toast';
import { eventBus } from '../../src/realtime/EventBus';

function renderOpen(onClose = () => {}) {
  return render(
    <ToastProvider>
      <DialogueUI open={true} onClose={onClose} agentId="a-1" />
    </ToastProvider>,
  );
}

describe('DialogueUI', () => {
  it('renders open state and closes on overlay click', () => {
    const onClose = vi.fn();
    render(
      <ToastProvider>
        <DialogueUI open={true} onClose={onClose} />
      </ToastProvider>,
    );
    const overlay = screen.getAllByTestId('dialogue-overlay')[0];
    fireEvent.click(overlay);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it.skip('switches tabs via keyboard shortcuts (1/3)', () => {
    renderOpen();
    // default tab is Thread, switch to Info and back
    fireEvent.keyDown(window, { key: '3' });
    expect(screen.getAllByRole('button', { name: /info tab/i })[0]).toHaveAttribute(
      'aria-pressed',
      'true',
    );
    fireEvent.keyDown(window, { key: '1' });
    expect(screen.getByRole('button', { name: /thread tab/i })).toHaveAttribute(
      'aria-pressed',
      'true',
    );
  });

  it.skip('auto-scrolls thread list when new messages arrive', async () => {
    renderOpen();
    // Ensure Thread tab active
    fireEvent.keyDown(window, { key: '1' });
    const list = (await screen.findAllByTestId('thread-list'))[0] as HTMLDivElement;
    // Prime dimensions
    Object.defineProperty(list, 'scrollHeight', { value: 500, configurable: true });
    Object.defineProperty(list, 'clientHeight', { value: 200, configurable: true });
    // Dispatch two work_stream messages
    eventBus.emit('work_stream', { agentId: 'a-1', message: 'hello' });
    eventBus.emit('work_stream', { agentId: 'a-1', message: 'world' });
    // Let RAF flush
    await new Promise((r) => setTimeout(r, 10));
    // Verify messages appended (auto-scroll behavior is implicitly exercised)
    expect(await screen.findByText(/hello/i)).toBeInTheDocument();
    expect(await screen.findByText(/world/i)).toBeInTheDocument();
  });

  it('Close button focuses on open and overlay has proper aria', () => {
    renderOpen();
    const overlay = screen.getAllByTestId('dialogue-overlay')[0];
    expect(overlay).toHaveAttribute('aria-hidden', 'false');
    const close = screen.getAllByRole('button', { name: /close dialogue/i })[0];
    // focus should move to Close shortly after mount; emulate timers
    close.focus();
    expect(document.activeElement).toBe(close);
  });

  it('matches snapshot in open state', () => {
    const { asFragment } = render(
      <ToastProvider>
        <DialogueUI open={true} onClose={() => {}} agentId="a-1" />
      </ToastProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
