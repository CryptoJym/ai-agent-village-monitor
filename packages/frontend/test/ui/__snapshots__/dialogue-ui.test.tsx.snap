// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DialogueUI > matches snapshot in open state 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    data-testid="dialogue-overlay"
    style="position: fixed; inset: 0; background: rgba(0, 0, 0, 0.35); pointer-events: auto; transition: background 200ms ease; z-index: 30;"
  >
    <section
      aria-label="Agent dialogue panel"
      aria-modal="true"
      data-testid="dialogue-panel"
      role="dialog"
      style="position: absolute; left: 0px; right: 0px; bottom: 0px; height: 30vh; background: rgb(11, 18, 32); border-top: 1px solid rgb(31, 41, 55); box-shadow: 0 -10px 30px rgba(0,0,0,0.25); transform: translateY(0%); transition: transform 300ms ease-out; display: flex; flex-direction: column;"
    >
      <header
        aria-label="Dialogue tabs"
        data-testid="dialogue-tabs"
        style="display: flex; gap: 8px; padding: 8px 12px; border-bottom: 1px solid rgb(31, 41, 55); background: rgb(15, 23, 42);"
      >
        <button
          aria-label="Thread tab (1)"
          aria-pressed="true"
          style="padding: 6px 10px; border-radius: 6px; background: rgb(31, 41, 55); color: rgb(229, 231, 235); border: 1px solid rgb(55, 65, 81); cursor: pointer;"
        >
          Thread
        </button>
        <button
          aria-label="Control tab (2)"
          aria-pressed="false"
          style="padding: 6px 10px; border-radius: 6px; background: transparent; color: rgb(229, 231, 235); border: 1px solid rgb(55, 65, 81); cursor: pointer;"
        >
          Control
        </button>
        <button
          aria-label="Info tab (3)"
          aria-pressed="false"
          style="padding: 6px 10px; border-radius: 6px; background: transparent; color: rgb(229, 231, 235); border: 1px solid rgb(55, 65, 81); cursor: pointer;"
        >
          Info
        </button>
        <div
          style="margin-left: auto;"
        >
          <button
            aria-label="Close dialogue (Esc)"
            style="padding: 6px 10px; border-radius: 6px; background: transparent; color: rgb(229, 231, 235); border: 1px solid rgb(55, 65, 81); cursor: pointer;"
          >
            Close (Esc)
          </button>
        </div>
      </header>
      <div
        style="flex: 1 1 0%; min-height: 0; overflow: hidden;"
      >
        <div
          style="display: flex; flex-direction: column; height: 100%;"
        >
          <div
            style="display: flex; align-items: center; justify-content: space-between; padding: 8px 12px; border-bottom: 1px solid rgb(31, 41, 55); background: rgb(11, 18, 32);"
          >
            <div
              style="color: rgb(148, 163, 184);"
            >
              Agent: a-1
            </div>
            <div
              data-testid="thread-status"
              style="display: flex; align-items: center; gap: 8px;"
            >
              <span
                style="display: inline-block; width: 8px; height: 8px; border-radius: 9999px; background: rgb(239, 68, 68);"
              />
              <span
                style="color: rgb(203, 213, 225);"
              >
                Disconnected
              </span>
            </div>
          </div>
          <div
            data-testid="thread-list"
            style="flex: 1 1 0%; overflow-y: auto; padding: 8px 12px; background: rgb(15, 23, 42);"
          />
          <div
            style="display: flex; gap: 8px; padding: 12px; border-top: 1px solid rgb(31, 41, 55); background: rgb(11, 18, 32);"
          >
            <input
              data-testid="thread-input"
              placeholder="Ask the agent..."
              style="flex: 1 1 0%; padding: 10px 12px; background: rgb(15, 23, 42); color: rgb(229, 231, 235); border: 1px solid rgb(55, 65, 81); border-radius: 8px; outline: none;"
              value=""
            />
            <button
              data-testid="thread-send"
              style="padding: 10px 14px; background: rgb(31, 41, 55); color: rgb(229, 231, 235); border: 1px solid rgb(55, 65, 81); border-radius: 8px; cursor: pointer;"
              type="button"
            >
              Send
            </button>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
